pkg/kubelet/kuberuntime/logs/logs_test.go

```diff
@@ -210,6 +212,144 @@ func TestReadLogs(t *testing.T) {
+	defer os.Remove(fileName)
+
+	file.WriteString(`{"log":"line1\n","stream":"stdout","time":"2023-02-14T11:18:01.00000000Z"}` + "\n")
+	file.WriteString(`{"log":"line2\n","stream":"stdout","time":"2023-02-14T11:18:02.00000000Z"}` + "\n")
+	file.WriteString(`{"log":"line3\n","stream":"stdout","time":"2023-02-14T11:18:03.00000000Z"}` + "\n")
+	// Rotate marker
+	file.WriteString(`{"log":"__ROTATE_MARKER__\n","stream":"stderr","time":"2023-02-14T11:18:04.00000000Z"}` + "\n")
+
+	podLogOptions := v1.PodLogOptions{
+		Follow: true,
+	}
+
+	containerID := "fake-container-id"
+	fakeRuntimeService := &apitesting.FakeRuntimeService{
+		Containers: map[string]*apitesting.FakeContainer{
+			containerID: {
+				ContainerStatus: runtimeapi.ContainerStatus{
+					State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+				},
+			},
+		},
+	}
+
+	opts := NewLogOptions(&podLogOptions, time.Now())
+	stdoutBuf := &syncBuffer{}
+	stderrBuf := &syncBuffer{}
+
+	expected := "line1\nline2\nline3\nline4\nline5\n"
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	evaluated, err := filepath.EvalSymlinks(file.Name())
+	if err != nil {
+		t.Fatalf("failed to try resolving symlinks in path %q: %v", file.Name(), err)
+	}
+
+	dir := filepath.Dir(evaluated)
+	baseName := filepath.Base(evaluated)
+
+	stopped := make(chan struct{})
+	go func() {
+		timeout := time.NewTimer(30 * time.Second)
+		ticker := time.NewTicker(10 * time.Millisecond)
+		rotated := false
+		defer cancel()
+		for {
+			select {
+			case <-ticker.C:
+				if !rotated {
+					if stderrBuf.String() == "__ROTATE_MARKER__\n" {
+						// Pretend to rotate the log.
+						rotatedName := fmt.Sprintf("%s.%s", baseName, time.Now().Format("220060102-150405"))
+						rotatedName = filepath.Join(dir, rotatedName)
+						if err := os.Rename(filepath.Join(dir, baseName), rotatedName); err != nil {
+							fmt.Printf("failed to rotate log %q to %q: %v", fileName, rotatedName, err)
+							return
+						}
+						os.Remove(rotatedName)
+						file.Close()
+
+						newF := filepath.Join(dir, baseName)
+						if f, err := os.Create(newF); err != nil {
+							fmt.Printf("unable to create the file")
+							return
+						} else {
+							// Write new logs to log file
+							f.WriteString(`{"log":"line4\n","stream":"stdout","time":"2023-02-14T11:18:05.00000000Z"}` + "\n")
+							f.WriteString(`{"log":"line5\n","stream":"stdout","time":"2023-02-14T11:18:06.00000000Z"}` + "\n")
+							f.Close()
+						}
+						rotated = true
+					}
+				} else {
+					if stdoutBuf.String() == expected {
+						close(stopped)
+						return
+					}
+				}
+			case <-timeout.C:
+				// Read logs timed out.
+				cancel()
+				return
+			}
+		}
+	}()
+
+	go func(ctx context.Context) {
+		if err = ReadLogs(ctx, file.Name(), containerID, opts, fakeRuntimeService, stdoutBuf, stderrBuf); err != nil {
+			if err.Error() != "context cancelled" {
+				fmt.Printf("readlog failed, err %v", err)
+				return
+			}
+		}
```

Comment:
Would it make sense to check if the context of stdoutBuf and stderrBuf matches to the test content we wrote?